main.py

```diff
@@ -109,38 +113,59 @@ def aws_sts_assume_role(
     if web_identity_token != "":
         arg_list += ["--web-identity-token", web_identity_token]
 
-    # set `AWS_EC2_METADATA_DISABLED` to avoid AWS CLI reaching out to metadata endpoint
-    # on GitHub-hosted runners, which causes runtime error
+    # setting `AWS_EC2_METADATA_DISABLED` stops the AWS CLI from reaching out
+    # to (a non-existent) metadata endpoint on GitHub hosted runners
     env_var_collection["AWS_EC2_METADATA_DISABLED"] = "true"
     env_var_collection["PATH"] = os.environ.get("PATH", "")
 
-    # execute AWS CLI command
-    try:
-        result = subprocess.run(
-            arg_list,
-            encoding="utf-8",
-            env=env_var_collection,
-            stderr=subprocess.PIPE,
-            stdout=subprocess.PIPE,
-        )
-    except FileNotFoundError as ex:
-        exit_error("unable to assume role, AWS CLI installed?")
-
-    if result.returncode != 0:
-        exit_error("unable to assume role: \n" + result.stderr.strip())
+    retry_remain = ASSUME_ROLE_RETRY_COUNT
+    retry_backoff_milliseconds = 0
+    result_stdout = ""
+    while True:
+        # execute AWS CLI command
+        retry_remain -= 1
+        try:
+            result = subprocess.run(
+                arg_list,
+                encoding="utf-8",
+                env=env_var_collection,
+                stderr=subprocess.PIPE,
+                stdout=subprocess.PIPE,
+            )
+        except FileNotFoundError as ex:
+            exit_error("unable to assume role, AWS CLI installed?")
+
+        if result.returncode == 0:
+            # hold result of successful execution, exit retry loop
+            result_stdout = result.stdout
+            break
+
+        # command execution resulted in error
+        result_stderr = result.stderr.strip()
+        if (retry_remain > 0) and retry_error_match_list:
+            # if returned error text contains item in error match list - retry
+            allow_retry = False
+            for item in retry_error_match_list:
+                if item in result_stderr:
+                    allow_retry = True
```

Comment:
Perhaps move this  logic into `def should_allow_retry(stderr)`? Getting a bit deep here.

Reply:
Great call @olliebun - updated and rebased into a `allow_retry()` - much nicer to read. 👍